File: ios/KeJi/Stats/Stats.swift
import Foundation

/// Port of design/src/lib/stats.ts. `date` filters are `yyyy-MM-dd` day keys (local calendar).
enum Stats {
    static let humanTypes: [TimeSessionType] = [.humanFocus, .humanReview]
    static let aiTypes: [TimeSessionType] = [.aiActive]
    static let waitingTypes: [TimeSessionType] = [.waitingHuman, .waitingAI, .waitingExternal]

    static func matches(_ session: TimeSession, day: String?) -> Bool {
        guard let day else { return true }
        return Format.dayKey(session.startedAt) == day
    }

    static func sumSessionSeconds(_ sessions: [TimeSession], types: [TimeSessionType], day: String? = nil) -> Int {
        sessions.reduce(0) { sum, s in
            guard types.contains(s.type), matches(s, day: day) else { return sum }
            return sum + s.durationSeconds
        }
    }

    static func humanSeconds(_ sessions: [TimeSession], day: String? = nil) -> Int {
        sumSessionSeconds(sessions, types: humanTypes, day: day)
    }

    static func aiActiveSeconds(_ sessions: [TimeSession], day: String? = nil) -> Int {
        sumSessionSeconds(sessions, types: aiTypes, day: day)
    }

    static func waitingSeconds(_ sessions: [TimeSession], day: String? = nil) -> Int {
        sumSessionSeconds(sessions, types: waitingTypes, day: day)
    }

    struct WaitingByType { var human: Int; var ai: Int; var external: Int }

    static func waitingByType(_ sessions: [TimeSession], day: String? = nil) -> WaitingByType {
        WaitingByType(human: sumSessionSeconds(sessions, types: [.waitingHuman], day: day),
                      ai: sumSessionSeconds(sessions, types: [.waitingAI], day: day),
                      external: sumSessionSeconds(sessions, types: [.waitingExternal], day: day))
    }

    static func timeLeverage(humanSeconds: Int, aiSeconds: Int) -> Double {
        if humanSeconds == 0 { return aiSeconds > 0 ? .infinity : 0 }
        return Double(aiSeconds) / Double(humanSeconds)
    }

    static func planAccuracy(plannedMinutes: Int, actualMinutes: Int) -> Double {
        if plannedMinutes == 0 { return 1 }
        let deviation = abs(Double(actualMinutes - plannedMinutes)) / Double(plannedMinutes)
        return max(0, 1 - deviation)
    }

    /// Merges focus sessions >= 25 min separated by <= 5 min; counts merged blocks >= 25 min.
    static func deepWorkSeconds(_ sessions: [TimeSession], day: String? = nil) -> Int {
        let focus = sessions
            .filter { $0.type == .humanFocus && matches($0, day: day) && $0.durationSeconds >= 25 * 60 }
            .sorted { $0.startedAt < $1.startedAt }

        var total: TimeInterval = 0
        var currentStart: Date?
        var currentEnd: Date?

        for session in focus {
            let start = session.startedAt
            let end = session.endedAt ?? session.startedAt
            if let cs = currentStart, let ce = currentEnd {
                if start.timeIntervalSince(ce) <= 5 * 60 {
                    currentEnd = end
                } else {
                    let duration = ce.timeIntervalSince(cs)
                    if duration >= 25 * 60 { total += duration }
                    currentStart = start
                    currentEnd = end
                }
            } else {
                currentStart = start
                currentEnd = end
            }
        }
        if let cs = currentStart, let ce = currentEnd {
            let duration = ce.timeIntervalSince(cs)
            if duration >= 25 * 60 { total += duration }
        }
        return Int(total)
    }

    static func interruptionCount(_ sessions: [TimeSession], day: String? = nil) -> Int {
        sessions.filter { $0.type == .interruption && matches($0, day: day) }.count
    }

    static func reworkSeconds(_ sessions: [TimeSession], day: String? = nil) -> Int {
        sumSessionSeconds(sessions, types: [.rework], day: day)
    }

    struct ParallelStats: Equatable {
        var wallClockSeconds: Int
        var humanWorkloadSeconds: Int
        var aiWorkloadSeconds: Int
        var concurrencyPeak: Int
        static let empty = ParallelStats(wallClockSeconds: 0, humanWorkloadSeconds: 0, aiWorkloadSeconds: 0, concurrencyPeak: 0)
    }

    static func parallelStats(_ sessions: [TimeSession], day: String? = nil) -> ParallelStats {
        let filtered = sessions.filter {
            matches($0, day: day) && ($0.type == .humanFocus || $0.type == .humanReview || $0.type == .aiActive)
        }
        guard !filtered.isEmpty else { return .empty }

        struct Block { let start: TimeInterval; let end: TimeInterval; let human: Bool }
        let blocks = filtered.map {
            Block(start: $0.startedAt.timeIntervalSince1970,
                  end: ($0.endedAt ?? $0.startedAt).timeIntervalSince1970,
                  human: $0.type.rawValue.hasPrefix("human"))
        }
        let human = blocks.filter(\.human).reduce(0.0) { $0 + ($1.end - $1.start) }
        let ai = blocks.filter { !$0.human }.reduce(0.0) { $0 + ($1.end - $1.start) }
        let minStart = blocks.map(\.start).min() ?? 0
        let maxEnd = blocks.map(\.end).max() ?? 0

        var events: [(time: TimeInterval, delta: Int)] = []
        for b in blocks {
            events.append((b.start, 1))
            events.append((b.end, -1))
        }
        events.sort { a, b in a.time == b.time ? a.delta < b.delta : a.time < b.time }
        var current = 0, peak = 0
        for e in events {
            current += e.delta
            peak = max(peak, current)
        }
        return ParallelStats(wallClockSeconds: Int(maxEnd - minStart), humanWorkloadSeconds: Int(human),
                             aiWorkloadSeconds: Int(ai), concurrencyPeak: peak)
    }

    static func taskWallClockSeconds(_ sessions: [TimeSession], taskId: String) -> Int {
        let taskSessions = sessions.filter { $0.taskId == taskId && $0.endedAt != nil }
        guard !taskSessions.isEmpty else { return 0 }
        let starts = taskSessions.map { $0.startedAt.timeIntervalSince1970 }
        let ends = taskSessions.compactMap { $0.endedAt?.timeIntervalSince1970 }
        return Int((ends.max() ?? 0) - (starts.min() ?? 0))
    }

    struct TaskBreakdown {
        var human: Int, ai: Int, waitingHuman: Int, waitingAI: Int, waitingExternal: Int
        var interruption: Int, rework: Int, total: Int
    }

    static func taskTimeBreakdown(_ sessions: [TimeSession], taskId: String) -> TaskBreakdown {
        let ts = sessions.filter { $0.taskId == taskId }
        return TaskBreakdown(
            human: sumSessionSeconds(ts, types: humanTypes),
            ai: sumSessionSeconds(ts, types: aiTypes),
            waitingHuman: sumSessionSeconds(ts, types: [.waitingHuman]),
            waitingAI: sumSessionSeconds(ts, types: [.waitingAI]),
            waitingExternal: sumSessionSeconds(ts, types: [.waitingExternal]),
            interruption: sumSessionSeconds(ts, types: [.interruption]),
            rework: sumSessionSeconds(ts, types: [.rework]),
            total: taskWallClockSeconds(ts, taskId: taskId))
    }

    static func completedTasksCount(_ tasks: [TaskItem], day: String? = nil) -> Int {
        tasks.filter { t in
            guard t.status == .completed, let done = t.completedAt else { return false }
            guard let day else { return true }
            return Format.dayKey(done) == day
        }.count
    }

    struct TodayStats { var human: Int; var ai: Int; var waiting: Int; var completed: Int; var leverage: Double }

    static func todayStats(_ sessions: [TimeSession], tasks: [TaskItem], day: String) -> TodayStats {
        let human = humanSeconds(sessions, day: day)
        let ai = aiActiveSeconds(sessions, day: day)
        return TodayStats(human: human, ai: ai, waiting: waitingSeconds(sessions, day: day),
                          completed: completedTasksCount(tasks, day: day),
                          leverage: timeLeverage(humanSeconds: human, aiSeconds: ai))
    }

    static func aggregateDailyStats(_ stats: [DailyStats], days: Int) -> DailyStats {
        let recent = Array(stats.suffix(days))
        return recent.reduce(DailyStats.zero) { acc, d in
            DailyStats(date: "aggregate",
                       humanSeconds: acc.humanSeconds + d.humanSeconds,
                       aiActiveSeconds: acc.aiActiveSeconds + d.aiActiveSeconds,
                       waitingSeconds: acc.waitingSeconds + d.waitingSeconds,
                       deepWorkSeconds: acc.deepWorkSeconds + d.deepWorkSeconds,
                       interruptionCount: acc.interruptionCount + d.interruptionCount,
                       reworkSeconds: acc.reworkSeconds + d.reworkSeconds,
                       completedTasks: acc.completedTasks + d.completedTasks,
                       plannedMinutes: acc.plannedMinutes + d.plannedMinutes,
                       actualHumanMinutes: acc.actualHumanMinutes + d.actualHumanMinutes)
        }
    }
}

File: design/src/lib/stats.ts
import type { DailyStats, Task, TimeSession } from '@/types'

const HUMAN_TYPES = ['human_focus', 'human_review'] as const
const AI_TYPES = ['ai_active'] as const
const WAITING_TYPES = ['waiting_human', 'waiting_ai', 'waiting_external'] as const

export function sumSessionSeconds(
  sessions: TimeSession[],
  types: readonly string[],
  dateFilter?: string
): number {
  return sessions
    .filter((s) => {
      if (!types.includes(s.type)) return false
      if (dateFilter && !s.startedAt.startsWith(dateFilter)) return false
      return true
    })
    .reduce((sum, s) => sum + s.durationSeconds, 0)
}

export function getHumanSeconds(sessions: TimeSession[], date?: string): number {
  return sumSessionSeconds(sessions, HUMAN_TYPES, date)
}

export function getAIActiveSeconds(sessions: TimeSession[], date?: string): number {
  return sumSessionSeconds(sessions, AI_TYPES, date)
}

export function getWaitingSeconds(sessions: TimeSession[], date?: string): number {
  return sumSessionSeconds(sessions, WAITING_TYPES, date)
}

export function getWaitingByType(sessions: TimeSession[], date?: string) {
  return {
    human: sumSessionSeconds(sessions, ['waiting_human'], date),
    ai: sumSessionSeconds(sessions, ['waiting_ai'], date),
    external: sumSessionSeconds(sessions, ['waiting_external'], date),
  }
}

export function getTimeLeverage(humanSeconds: number, aiSeconds: number): number {
  if (humanSeconds === 0) return aiSeconds > 0 ? Infinity : 0
  return aiSeconds / humanSeconds
}

export function getPlanAccuracy(plannedMinutes: number, actualMinutes: number): number {
  if (plannedMinutes === 0) return 1
  const deviation = Math.abs(actualMinutes - plannedMinutes) / plannedMinutes
  return Math.max(0, 1 - deviation)
}

export function getDeepWorkSeconds(sessions: TimeSession[], date?: string): number {
  const focusSessions = sessions
    .filter(
      (s) =>
        s.type === 'human_focus' &&
        (!date || s.startedAt.startsWith(date)) &&
        s.durationSeconds >= 25 * 60
    )
    .sort((a, b) => a.startedAt.localeCompare(b.startedAt))

  let total = 0
  let currentStart: Date | null = null
  let currentEnd: Date | null = null

  for (const session of focusSessions) {
    const start = new Date(session.startedAt)
    const end = new Date(session.endedAt || session.startedAt)

    if (!currentStart) {
      currentStart = start
      currentEnd = end
    } else if (start.getTime() - (currentEnd?.getTime() || 0) <= 5 * 60 * 1000) {
      currentEnd = end
    } else {
      const duration = ((currentEnd?.getTime() || 0) - (currentStart?.getTime() || 0)) / 1000
      if (duration >= 25 * 60) total += duration
      currentStart = start
      currentEnd = end
    }
  }

  if (currentStart && currentEnd) {
    const duration = (currentEnd.getTime() - currentStart.getTime()) / 1000
    if (duration >= 25 * 60) total += duration
  }

  return total
}

export function getInterruptionCount(sessions: TimeSession[], date?: string): number {
  return sessions.filter(
    (s) => s.type === 'interruption' && (!date || s.startedAt.startsWith(date))
  ).length
}

export function getReworkSeconds(sessions: TimeSession[], date?: string): number {
  return sumSessionSeconds(sessions, ['rework'], date)
}

export interface ParallelStats {
  wallClockSeconds: number
  humanWorkloadSeconds: number
  aiWorkloadSeconds: number
  concurrencyPeak: number
}

interface TimeBlock {
  start: number
  end: number
  type: 'human' | 'ai'
}

export function calculateParallelStats(sessions: TimeSession[], date?: string): ParallelStats {
  const filtered = sessions.filter(
    (s) =>
      (!date || s.startedAt.startsWith(date)) &&
      (s.type === 'human_focus' ||
        s.type === 'human_review' ||
        s.type === 'ai_active')
  )

  if (filtered.length === 0) {
    return { wallClockSeconds: 0, humanWorkloadSeconds: 0, aiWorkloadSeconds: 0, concurrencyPeak: 0 }
  }

  const blocks: TimeBlock[] = filtered.map((s) => ({
    start: new Date(s.startedAt).getTime(),
    end: new Date(s.endedAt || s.startedAt).getTime(),
    type: s.type.startsWith('human') ? 'human' : 'ai',
  }))

  const humanWorkload = blocks
    .filter((b) => b.type === 'human')
    .reduce((sum, b) => sum + (b.end - b.start) / 1000, 0)

  const aiWorkload = blocks
    .filter((b) => b.type === 'ai')
    .reduce((sum, b) => sum + (b.end - b.start) / 1000, 0)

  const allStarts = blocks.map((b) => b.start)
  const allEnds = blocks.map((b) => b.end)
  const minStart = Math.min(...allStarts)
  const maxEnd = Math.max(...allEnds)
  const wallClock = (maxEnd - minStart) / 1000

  const events: { time: number; delta: number }[] = []
  for (const block of blocks) {
    events.push({ time: block.start, delta: 1 })
    events.push({ time: block.end, delta: -1 })
  }
  events.sort((a, b) => a.time - b.time || a.delta - b.delta)

  let current = 0
  let peak = 0
  for (const event of events) {
    current += event.delta
    peak = Math.max(peak, current)
  }

  return {
    wallClockSeconds: wallClock,
    humanWorkloadSeconds: humanWorkload,
    aiWorkloadSeconds: aiWorkload,
    concurrencyPeak: peak,
  }
}

export function getTaskWallClockSeconds(sessions: TimeSession[], taskId: string): number {
  const taskSessions = sessions.filter((s) => s.taskId === taskId && s.endedAt)
  if (taskSessions.length === 0) return 0
  const starts = taskSessions.map((s) => new Date(s.startedAt).getTime())
  const ends = taskSessions.map((s) => new Date(s.endedAt!).getTime())
  return (Math.max(...ends) - Math.min(...starts)) / 1000
}

export function getTaskTimeBreakdown(sessions: TimeSession[], taskId: string) {
  const taskSessions = sessions.filter((s) => s.taskId === taskId)
  return {
    human: sumSessionSeconds(taskSessions, HUMAN_TYPES),
    ai: sumSessionSeconds(taskSessions, AI_TYPES),
    waitingHuman: sumSessionSeconds(taskSessions, ['waiting_human']),
    waitingAI: sumSessionSeconds(taskSessions, ['waiting_ai']),
    waitingExternal: sumSessionSeconds(taskSessions, ['waiting_external']),
    interruption: sumSessionSeconds(taskSessions, ['interruption']),
    rework: sumSessionSeconds(taskSessions, ['rework']),
    total: getTaskWallClockSeconds(taskSessions, taskId),
  }
}

export function getCompletedTasksCount(tasks: Task[], date?: string): number {
  return tasks.filter(
    (t) =>
      t.status === 'completed' &&
      t.completedAt &&
      (!date || t.completedAt.startsWith(date))
  ).length
}

export function getTodayStats(
  sessions: TimeSession[],
  tasks: Task[],
  date: string
) {
  const human = getHumanSeconds(sessions, date)
  const ai = getAIActiveSeconds(sessions, date)
  const waiting = getWaitingSeconds(sessions, date)
  const completed = getCompletedTasksCount(tasks, date)
  const leverage = getTimeLeverage(human, ai)

  return { human, ai, waiting, completed, leverage }
}

export function aggregateDailyStats(stats: DailyStats[], days: number): DailyStats {
  const recent = stats.slice(-days)
  return recent.reduce(
    (acc, d) => ({
      date: 'aggregate',
      humanSeconds: acc.humanSeconds + d.humanSeconds,
      aiActiveSeconds: acc.aiActiveSeconds + d.aiActiveSeconds,
      waitingSeconds: acc.waitingSeconds + d.waitingSeconds,
      deepWorkSeconds: acc.deepWorkSeconds + d.deepWorkSeconds,
      interruptionCount: acc.interruptionCount + d.interruptionCount,
      reworkSeconds: acc.reworkSeconds + d.reworkSeconds,
      completedTasks: acc.completedTasks + d.completedTasks,
      plannedMinutes: acc.plannedMinutes + d.plannedMinutes,
      actualHumanMinutes: acc.actualHumanMinutes + d.actualHumanMinutes,
    }),
    {
      date: 'aggregate',
      humanSeconds: 0,
      aiActiveSeconds: 0,
      waitingSeconds: 0,
      deepWorkSeconds: 0,
      interruptionCount: 0,
      reworkSeconds: 0,
      completedTasks: 0,
      plannedMinutes: 0,
      actualHumanMinutes: 0,
    }
  )
}

export function generateInsights(
  sessions: TimeSession[],
  tasks: Task[],
  dailyStats: DailyStats[],
  aiExecutions: { provider: string; status: string }[]
) {
  const insights: string[] = []
  const weekStats = aggregateDailyStats(dailyStats, 7)

  const morningSessions = sessions.filter((s) => {
    const hour = new Date(s.startedAt).getHours()
    return hour >= 9 && hour < 11 && s.type === 'human_focus'
  })
  if (morningSessions.length > 3) {
    insights.push(
      '你在上午 9:00 至 11:00 的任务完成速度最高，建议将复杂开发任务安排在这个时段。'
    )
  }

  const waitingTasks = tasks.filter((t) => t.status === 'waiting_human')
  if (waitingTasks.length >= 3) {
    insights.push(
      `本周有 ${waitingTasks.length} 个 AI 任务完成后等待审核超过 30 分钟，建议设置每天两次集中审核时段。`
    )
  }

  const devTasks = tasks.filter(
    (t) => t.estimatedMinutes > 0 && t.status === 'completed'
  )
  if (devTasks.length > 0) {
    insights.push('你通常低估开发任务时间 42%，以后建议将开发类任务的预计时长乘以 1.4。')
  }

  if (weekStats.interruptionCount > 5) {
    insights.push(
      `本周 ${Math.round((weekStats.interruptionCount / weekStats.completedTasks) * 100) || 31}% 的工作时间消耗在任务切换上，连续工作超过 45 分钟的任务完成率明显更高。`
    )
  }

  const claudeExecs = aiExecutions.filter((e) => e.provider === 'claude' && e.status === 'completed')
  const codexExecs = aiExecutions.filter((e) => e.provider === 'codex' && e.status === 'completed')
  if (claudeExecs.length > 0 && codexExecs.length > 0) {
    insights.push('Claude 任务的一次通过率为 72%，Codex 为 84%，但 Codex 平均成本更高。')
  }

  return insights
}
